Initial implementation of WebExtensionSidebar external API

AriYoung00 opened the pull request:
#### aa35126f4291babb0cac50a4a3683dd5cfbc33b0
<pre>
Initial implementation of WebExtensionSidebar external API
<a href="https://webkit.org/b/277575">https://webkit.org/b/277575</a>
<a href="https://rdar.apple.com/133073385">rdar://133073385</a>

Reviewed by Timothy Hatcher.

This patch introduces an initial external API for the
WebExtensionSidebar object.

* Source/WebKit/Modules/OSX_Private.modulemap: Add
  _WKWebExtensionSidebar module
* Source/WebKit/Modules/iOS_Private.modulemap: Add
  _WKWebExtensionSidebar module
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject): Add case for Type::WebExtensionSidebar which
allocates a new _WKWebExtensionSidebar object
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
(-[WKWebExtensionContext sidebarForTab:]): Added method implementation
to retrieve the sidebar for the given tab from the appropriate
WebExtensionContext
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContextPrivate.h: Add
  prototype for [WKWebExtensionContext sidebarForTab:]
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
Add _webExtensionController presentSidebar to
WKWebExtensionControllerDelegatePrivate protocol.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.mm: Added.
(-[_WKWebExtensionSidebar webExtensionContext]): External getter for a
sidebar&apos;s context
(-[_WKWebExtensionSidebar title]): External getter for a sidebar&apos;s title
(-[_WKWebExtensionSidebar iconForSize:]): External getter for a
sidebar&apos;s icon
(-[_WKWebExtensionSidebar isEnabled]): External getter for a sidebar&apos;s
enablement state
(-[_WKWebExtensionSidebar webView]): External getter for a sidebar&apos;s web
view
(-[_WKWebExtensionSidebar willOpenSidebar]): Method to indicate the
sidebar will be opened
(-[_WKWebExtensionSidebar willCloseSidebar]): Method to indicate the
sidebar will be closed
(-[_WKWebExtensionSidebar associatedTab]): External getter for a
sidebar&apos;s tab
(-[_WKWebExtensionSidebar _apiObject]): Internal getter for the
sidebar&apos;s implementation object as a API::Object
(-[_WKWebExtensionSidebar _webExtensionSidebar]): Internal getter for
the sidebar&apos;s implementation object
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebarInternal.h:
  Added
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::hasSidebar): Utility method to check if the
extension uses either sidebar API
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
(toOptionalRef): Utility method to convert a RefPtr&lt;T&gt; to a
std::optional&lt;Ref&lt;T&gt;&gt;
(-[_WKWebExtensionSidebarWebViewDelegate initWithWebExtensionSidebar:]):
Initializer
(-[_WKWebExtensionSidebarWebViewDelegate webView:decidePolicyForNavigationAction:decisionHandler:]):
Delegate method implementation which handles deciding the navigation
policy for navigation actions within a sidebar&apos;s WKWebView
(WebKit::WebExtensionSidebar::setTitle): Notify delegate of title change
when necessary
(WebKit::WebExtensionSidebar::setSidebarPath): Notify delegate of
webView change when necessary
(WebKit::WebExtensionSidebar::willOpenSidebar): Assert that this sidebar
(or parent) has a web view
(WebKit::WebExtensionSidebar::willCloseSidebar): If this sidebar is a
tab-specific sidebar, and if this sidebar has its own web view, set the
web view to nil
(WebKit::WebExtensionSidebar::reloadWebView): Helper method to load the
current sidebar path in the sidebar&apos;s web view, if present
(WebKit::WebExtensionSidebar::webView): Use parent web view if no path
is set, properly configure web view before returning
(WebKit::WebExtensionSidebar::webViewWasUpdated): Helper method to
notify delegate to re-query this sidebar&apos;s web view
* Source/WebKit/UIProcess/Extensions/WebExtension.h: Add prototype for
  hasSidebar
* Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h: Add wrapper
  method, sidebar API wrapper prototypes, add declarations for
  willOpenSidebar and willCloseSidebar
(WebKit::WebExtensionSidebar::wrapper const): Added
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add new files to
  project

Canonical link: <a href="https://commits.webkit.org/282878@main">https://commits.webkit.org/282878@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e31e85295fddb2a9de9acd3947f8bded918a4801

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64547 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43912 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17142 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68568 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15153 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51659 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15432 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51932 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10454 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67613 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40623 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55856 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32553 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37289 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14026 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59228 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13565 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70267 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8492 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13070 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59254 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8526 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55945 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59428 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7026 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/699 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9788 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39723 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/40801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/41984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40545 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
```diff
@@ -1229,4 +1229,34 @@ - (NSURL *)_backgroundContentURL
+        .and_then([](Ref<WebKit::WebExtensionSidebar> const& sidebar) -> std::optional<_WKWebExtensionSidebar *> {
+            return (_WKWebExtensionSidebar *) sidebar->wrapper();
+        })
+        .value_or(nil);
```
All this logic should be in `WebExtensionContextCocoa` instead. The API object should be a thin wrapper around the C++ logic.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
```diff
@@ -1229,4 +1229,34 @@ - (NSURL *)_backgroundContentURL
+- (nullable _WKWebExtensionSidebar *)sidebarForTab:(id<WKWebExtensionTab>)tab
+{
+    if (!_webExtensionContext->extension().hasSidebar())
+        return nil;
```
Ditto for this.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContextPrivate.h:
```diff
@@ -35,6 +37,13 @@ WK_HEADER_AUDIT_BEGIN(nullability, sendability)
+/*!
+ @abstract Retrieves the extension sidebar for a given tab, or the default sidebar if `nil` is passed.
+ @param tab The tab for which to retrieve the extension sidebar, or `nil` to get the default sidebar.
+ @discussion The returned object represents the sidebar which should be displayed when the given tab is active. When the tab is not provided, it will return the default sidebar; however, this sidebar should not be directly displayed. Even if there are no tab-specific properties set for the current tab, the tab-specific sidebar object should always be requested here; this way, WebKit can update pr ...
```
Should hard wrap to a reasonable width. Something more aligned with action would be best.

```
@discussion The returned object represents the sidebar specific to the tab when provided; otherwise, it returns the default sidebar.
The default sidebar should not be directly displayed. When possible, specify the tab to get the most context-relevant sidebar.
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
```diff
@@ -67,6 +69,8 @@ WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA))
  */
 - (void)_webExtensionController:(WKWebExtensionController *)controller didCreateBackgroundWebView:(WKWebView *)webView forExtensionContext:(WKWebExtensionContext *)context;
 
+- (void)_webExtensionController:(WKWebExtensionController *)controller presentSidebar:(_WKWebExtensionSidebar *)sidebar forExtensionContext:(WKWebExtensionContext *)context completionHandler:(void (^)(NSError * _Nullable error))completionHandler;
```
Should add header doc here too so it is ready to go.

AriYoung00 replied:
Whoops, missed this one before I opened the PR -- adding now

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h:
```diff
@@ -0,0 +1,142 @@
+ */
+- (void)iconWasUpdated;
+
+@end
```
I wasn't expecting delegates per sidebar. Hmm. This is likely too granular.

I was imagining one or two methods on the existing extension controller delegate.

Something like:

```objc
/*!
 @abstract Called when the sidebar for a specific tab is updated.
 @param controller The web extension controller initiating the update.
 @param sidebar The sidebar object that has been updated.
 @param context The context within which the web extension is running.
 @discussion This method is called whenever the sidebar content is updated due to script actions or other triggers within the extension. 
 This can include changes to the sidebar's appearance, content, or state. Implementing this method is necessary if the app intends to 
 react to dynamic updates in the sidebar and ensure the UI is synchronized with the latest sidebar state.
 */
- (void)_webExtensionController:(WKWebExtensionController *)controller sidebarDidUpdate:(_WKWebExtensionSidebar *)sidebar forExtensionContext:(WKWebExtensionContext *)context;
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h:
```diff
@@ -0,0 +1,142 @@
+@property (nonatomic, readonly, getter=isEnabled) BOOL enabled;
+
+/*! @abstract The web view which should be displayed when this sidebar is opened. */
+@property (nonatomic, readonly, nullable) WKWebView *webView;
```
While this is good, we likely also want to expose a `{UI,NS}ViewController` that the app can add to the UI and that gives us control if the web view needs swapped out, etc. (Similar to how `WKWebExtensionAction` exposes a web view and view controller / popup.)

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h:
```diff
@@ -0,0 +1,142 @@
+ must query the sidebar in order to obtain a reference to this new ``WKWebView`` to display. This method is not necessarily called when the sidebar's panel
+ URL is updated: if the URL can be updated without replacing the ``WKWebView``, it will be.
+ */
+- (void)webViewWasUpdated;
```
What triggers a web view to change vs the URL just navigating the existing web view?

See comment about a view controller to help with this.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h:
```diff
@@ -0,0 +1,142 @@
+@property (nonatomic, nullable, weak) id <_WKWebExtensionActiveSidebarDelegate> activeSidebarDelegate;
+
+/*! @abstract The delegate for this sidebar which will receive callbacks when it is enabled, disabled, or the icon is changed. */
+@property (nonatomic, nullable, weak) id <_WKWebExtensionInactiveSidebarDelegate> inactiveSidebarDelegate;
```
A single delegate would be best, if not just use the controller delegate like I mention elsewhere.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h:
```diff
@@ -0,0 +1,142 @@
+ displayed. If this method is not called before the sidebar is opened, then the ``WKWebView`` associated with this sidebar may not have a
+ document loaded.
+ */
+- (void)willOpenSidebar;
+
+/*!
+ @abstract Indicate that the sidebar will be closed
+ @discussion This method should be invoked by the browser when the sidebar will be closed -- i.e., its associated ``WKWebView`` will cease
+ to be displayed. If this method is not called when the sidebar is closed, then the sidebar's associated ``WKWebView`` may remain active longer than
+ necessary. Note that calling this method does not guarantee that the ``WKWebView`` associated with a particular sidebar will be deallocated, as the
+ web view may be shared between mutliple sidebars.
+ */
+- (void)willCloseSidebar;
```
A view controller will help with this. But `WKWebExtensionAction` loads the web view on first access, and there is a `closePopup` method to close it. I would suggest similar here.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h:
```diff
@@ -0,0 +1,142 @@
+ @abstract Get the sidebar icon of the given size.
+ @discussion If the sidebar does not specify an icon, this will instead return the extension's action icon.
+ */
+- (NSImage *)iconForSize:(CGSize)size;
```
This would be better grouped up near `title`.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.mm:
```diff
@@ -0,0 +1,141 @@
+    return nil;
+}
+
+
```
```suggestion

```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -1123,6 +1123,10 @@ static WKWebExtensionError toAPI(WebExtension::Error error)
+{
+    return objectForKey<NSDictionary>(m_manifest, sidebarActionManifestKey) || hasRequestedPermission(WKWebExtensionPermissionSidePanel);
+}
 CocoaImage *WebExtension::sidebarIcon(CGSize idealSize)
```
```suggestion
}

CocoaImage *WebExtension::sidebarIcon(CGSize idealSize)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -215,19 +304,86 @@
+
+    // If we have no particular path configured for this sidebar, just give the parent's WebView (whose path will apply to this sidebar)
+    if (std::optional<Ref<WebExtensionSidebar>> parentRef; !m_sidebarPathOverride && !isDefaultSidebar() && (parentRef = parent()))
+        return parentRef.value()->webView();
```
AH, I think I see now. But we should just always own the web view per sidebar. Sharing a web view will lead to more problems than it is worth.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -238,6 +238,7 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
     bool hasPageAction();
 
 #if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+    bool hasSidebar();
```
You should consider exposing this on `WKWebExtension` too, with an underscore in `WKWebExtensionPrivate.h` for now.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
```diff
@@ -67,6 +69,17 @@ WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA))
+ @param controller The web extension controller initiating the request
+ @param sidebar The sidebar which should be displayed
+ @param context The context within which the web extension is running
+ @param completionHandler A block to be called once the sidebar has been opened
```
```suggestion
 @abstract Called when a sidebar is requested to be opened.
 @param controller The web extension controller initiating the request.
 @param sidebar The sidebar which should be displayed.
 @param context The context within which the web extension is running.
 @param completionHandler A block to be called once the sidebar has been opened.
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPrivate.h:
```diff
@@ -82,6 +82,9 @@ WK_HEADER_AUDIT_BEGIN(nullability, sendability)
 /*! @abstract A Boolean value indicating whether the extension use modules for the background content. */
 @property (readonly, nonatomic) BOOL _hasModularBackgroundContent;
 
+/*! @abstract A Boolean value indicating whether the extension uses extension sidebars. */
```
```suggestion
/*! @abstract A Boolean value indicating whether the extension has a sidebar. */
```

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
```diff
@@ -1229,4 +1229,25 @@ - (NSURL *)_backgroundContentURL
 #endif // ENABLE(WK_WEB_EXTENSIONS)
 
+#if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+
```
nit: can drop this newline

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
```diff
@@ -1229,4 +1229,25 @@ - (NSURL *)_backgroundContentURL
+        return maybeSidebar->wrapper();
+    return nil;
+}
+
```
same for 1243 and 1245

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
```diff
@@ -510,4 +510,16 @@ - (BOOL)_hasModularBackgroundContent
 
 #endif // ENABLE(WK_WEB_EXTENSIONS)
 
+#if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
```
would it be better to move this `ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)` case under `ENABLE(WK_WEB_EXTENSIONS)` and the `!ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)` under `!ENABLE(WK_WEB_EXTENSIONS)`?

AriYoung00 replied:
I generally keep them separate because it looks a tiny bit cleaner, and because `ENABLE_WK_WEB_EXTENSIONS_SIDEBAR` depends on `ENABLE_WK_WEB_EXTENSIONS`:

from `PlatformEnableCocoa.h`: `#define ENABLE_WK_WEB_EXTENSIONS_SIDEBAR 0 && ENABLE_WK_WEB_EXTENSIONS`

kiaraarose replied:
I imagine we would eventually get rid of `ENABLE_WK_WEB_EXTENSIONS_SIDEBAR` when it ships (or at least this is what we do in Safari, not sure if that's the approach we take in WebKit), which is why I think it'd be better to move it within the `ENABLE(WK_WEB_EXTENSIONS)` block

AriYoung00 replied:
Alrighty, just moved this around (and another similar occurrence as well)

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
```diff
@@ -510,4 +510,16 @@ - (BOOL)_hasModularBackgroundContent
 
 #endif // ENABLE(WK_WEB_EXTENSIONS)
```
should be `!ENABLE(WK_WEB_EXTENSIONS)`

AriYoung00 replied:
Sorry, not sure I'm understanding -- isn't the initial condition `#if ENABLE(WK_WEB_EXTENSIONS`? Why would we want to invert the condition on the closing `#endif`?

kiaraarose replied:
Usually we do something like `#endif // !ENABLE(WK_WEB_EXTENSIONS)` if we have an `#if`, `#else`, `#endif` to indicate that the second block is for the `not` case. but actually, i think because there's the `// ENABLE(WK_WEB_EXTENSIONS)` with the `#else` block, `ENABLE(WK_WEB_EXTENSIONS)` is actually correct, based on what i'm seeing in the codebase. so you can disregard my first comment.

kiaraarose commented on Source/WebKit/Shared/Cocoa/APIObject.mm:
```diff
@@ -437,6 +438,12 @@
+        wrapper = [_WKWebExtensionSidebar alloc];
+        break;
+#endif // ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
 #endif
```
i think it would be better to actually add this for the `endif` below
`#endif // ENABLE(WK_WEB_EXTENSIONS)`

the line enwrapping `ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)` are pretty short so i don't really think it's needed

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h:
```diff
@@ -0,0 +1,153 @@
+WK_HEADER_AUDIT_BEGIN(nullability, sendability)
+
+/*!
+ @abstract This delegate notifies the browser that properties relevant to a currently-open sidebar have changed,
```
should be `.` and then end, instead of `,`

AriYoung00 replied:
This delegate is gone now so shouldn't need to update this

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h:
```diff
@@ -0,0 +1,153 @@
+WK_HEADER_AUDIT_BEGIN(nullability, sendability)
+
+/*!
+ @abstract This delegate notifies the browser that properties relevant to a currently-open sidebar have changed,
```
`currently-open`, is the `-` needed here?

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h:
```diff
@@ -0,0 +1,153 @@
+@property (nonatomic, readonly, getter=isEnabled) BOOL enabled;
+
+/*! @abstract The web view which should be displayed when this sidebar is opened. */
+@property (nonatomic, readonly, nullable) WKWebView *webView;
```
nit: `nullable` before `nonatomic` per style guide
`@property (nonatomic, nullable, readonly) WKWebView *webView;`

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.mm:
```diff
@@ -0,0 +1,158 @@
+    return nil;
+}
+
+#endif // ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
```
`#endif // !ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)`

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -33,20 +33,103 @@
+        decisionHandler(WKNavigationActionPolicyCancel);
+        return;
+    }
+
```
drop extra newline

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -33,20 +33,146 @@
+    return self;
+}
+
+- (void) viewWillAppear
```
```suggestion
- (void)viewWillAppear
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -33,20 +33,146 @@
+        _webExtensionSidebar->willOpenSidebar();
+}
+
+- (void) viewWillDisappear
```
```suggestion
- (void)viewWillDisappear
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -220,14 +379,146 @@
+    if (!delegate)
+        return;
+
+    if ([delegate respondsToSelector:@selector(_webExtensionController:didUpdateSidebar:forExtensionContext:)]) {
```
```suggestion
    auto *delegate = extensionController->delegate();
    if (![delegate respondsToSelector:@selector(_webExtensionController:didUpdateSidebar:forExtensionContext:)])
        return;
```

AriYoung00 committed:
Initial implementation of WebExtensionSidebar external API
https://webkit.org/b/277575
rdar://133073385

Reviewed by Timothy Hatcher.

This patch introduces an initial external API for the
WebExtensionSidebar object.

* Source/WebKit/Modules/OSX_Private.modulemap: Add
  _WKWebExtensionSidebar module
* Source/WebKit/Modules/iOS_Private.modulemap: Add
  _WKWebExtensionSidebar module
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject): Add case for Type::WebExtensionSidebar which
allocates a new _WKWebExtensionSidebar object
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContext.mm:
(-[WKWebExtensionContext sidebarForTab:]): Added method implementation
to retrieve the sidebar for the given tab from the appropriate
WebExtensionContext
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionContextPrivate.h: Add
  prototype for [WKWebExtensionContext sidebarForTab:]
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
Add _webExtensionController presentSidebar to
WKWebExtensionControllerDelegatePrivate protocol.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebar.mm: Added.
(-[_WKWebExtensionSidebar webExtensionContext]): External getter for a
sidebar's context
(-[_WKWebExtensionSidebar title]): External getter for a sidebar's title
(-[_WKWebExtensionSidebar iconForSize:]): External getter for a
sidebar's icon
(-[_WKWebExtensionSidebar isEnabled]): External getter for a sidebar's
enablement state
(-[_WKWebExtensionSidebar webView]): External getter for a sidebar's web
view
(-[_WKWebExtensionSidebar willOpenSidebar]): Method to indicate the
sidebar will be opened
(-[_WKWebExtensionSidebar willCloseSidebar]): Method to indicate the
sidebar will be closed
(-[_WKWebExtensionSidebar associatedTab]): External getter for a
sidebar's tab
(-[_WKWebExtensionSidebar _apiObject]): Internal getter for the
sidebar's implementation object as a API::Object
(-[_WKWebExtensionSidebar _webExtensionSidebar]): Internal getter for
the sidebar's implementation object
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionSidebarInternal.h:
  Added
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::hasSidebar): Utility method to check if the
extension uses either sidebar API
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
(toOptionalRef): Utility method to convert a RefPtr<T> to a
std::optional<Ref<T>>
(-[_WKWebExtensionSidebarWebViewDelegate initWithWebExtensionSidebar:]):
Initializer
(-[_WKWebExtensionSidebarWebViewDelegate webView:decidePolicyForNavigationAction:decisionHandler:]):
Delegate method implementation which handles deciding the navigation
policy for navigation actions within a sidebar's WKWebView
(WebKit::WebExtensionSidebar::setTitle): Notify delegate of title change
when necessary
(WebKit::WebExtensionSidebar::setSidebarPath): Notify delegate of
webView change when necessary
(WebKit::WebExtensionSidebar::willOpenSidebar): Assert that this sidebar
(or parent) has a web view
(WebKit::WebExtensionSidebar::willCloseSidebar): If this sidebar is a
tab-specific sidebar, and if this sidebar has its own web view, set the
web view to nil
(WebKit::WebExtensionSidebar::reloadWebView): Helper method to load the
current sidebar path in the sidebar's web view, if present
(WebKit::WebExtensionSidebar::webView): Use parent web view if no path
is set, properly configure web view before returning
(WebKit::WebExtensionSidebar::webViewWasUpdated): Helper method to
notify delegate to re-query this sidebar's web view
* Source/WebKit/UIProcess/Extensions/WebExtension.h: Add prototype for
  hasSidebar
* Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h: Add wrapper
  method, sidebar API wrapper prototypes, add declarations for
  willOpenSidebar and willCloseSidebar
(WebKit::WebExtensionSidebar::wrapper const): Added
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add new files to
  project

Canonical link: https://commits.webkit.org/282878@main
